File: examples/canonical-analytics/python/src/analytics_workbench/charts/incident_bar.py
import plotly.graph_objects as go
from ..analytics.incidents import IncidentSeverity
from datetime import datetime

def build_incident_bar_figure(agg: IncidentSeverity) -> dict:
    fig = go.Figure()
    
    total_count = sum(agg.counts) if agg.counts else 0
    
    if not agg.severities:
        fig.add_annotation(text="No data available", showarrow=False, font=dict(size=20))
        fig.update_layout(title_text="Incident Count by Severity", xaxis_title_text="Severity", yaxis_title_text="Count")
    else:
        fig.add_trace(go.Bar(
            x=agg.severities,
            y=agg.counts,
            name="Incidents",
            hovertemplate="Severity: %{x}<br>Count: %{y}<extra></extra>"
        ))
        
        fig.update_layout(
            title_text="Incident Count by Severity",
            xaxis_title_text="Severity",
            yaxis_title_text="Count"
        )
        
    fig_dict = fig.to_dict()
    fig_dict["meta"] = {
        "total_count": total_count,
        "filters_applied": [],
        "generated_at": datetime.utcnow().isoformat()
    }
    return fig_dict

File: examples/canonical-analytics/rust/src/charts/incident_bar.rs
use super::to_figure_json;
use crate::analytics::IncidentSeverity;
use plotly::common::Title;
use plotly::{Bar, Plot};
use serde_json::Value;

pub fn build_incident_bar_figure(agg: &IncidentSeverity) -> Value {
    let mut plot = Plot::new();

    if agg.severities.is_empty() {
        let trace = Bar::new(Vec::<String>::new(), Vec::<i64>::new());
        plot.add_trace(trace);
        let layout = plotly::Layout::new().title(Title::with_text("No data for selected filters"));
        plot.set_layout(layout);
        return to_figure_json(&plot, 0, true);
    }

    let total_count: i64 = agg.counts.iter().sum();

    let trace = Bar::new(agg.severities.clone(), agg.counts.clone())
        .hover_template("Severity: %{x}<br>Count: %{y}<extra></extra>");
    plot.add_trace(trace);

    let layout = plotly::Layout::new()
        .title(Title::with_text("Incidents by Severity"))
        .x_axis(plotly::layout::Axis::new().title(Title::with_text("Severity")))
        .y_axis(plotly::layout::Axis::new().title(Title::with_text("Count")));
    plot.set_layout(layout);

    to_figure_json(&plot, total_count, true)
}
